command/apply.go

```diff
@@ -144,103 +206,75 @@ func (c *ApplyCommand) Run(args []string) int {
+	var diags tfdiags.Diagnostics
 
 	// Applying changes with dev overrides in effect could make it impossible
 	// to switch back to a release version if the schema isn't compatible,
 	// so we'll warn about it.
 	diags = diags.Append(c.providerDevOverrideRuntimeWarnings())
 
-	// Before we delegate to the backend, we'll print any warning diagnostics
-	// we've accumulated here, since the backend will start fresh with its own
-	// diagnostics.
-	c.showDiagnostics(diags)
-	diags = nil
-
 	// Build the operation
 	opReq := c.Operation(be)
-	opReq.AutoApprove = autoApprove
-	opReq.ConfigDir = configPath
+	opReq.AutoApprove = args.AutoApprove
+	opReq.ConfigDir = "."
 	opReq.Destroy = c.Destroy
-	opReq.Hooks = []terraform.Hook{countHook, c.uiHook()}
+	opReq.Hooks = view.Hooks()
 	opReq.PlanFile = planFile
-	opReq.PlanRefresh = refresh
-	opReq.ShowDiagnostics = c.showDiagnostics
+	opReq.PlanRefresh = args.Refresh
+	opReq.Targets = args.Targets
 	opReq.Type = backend.OperationTypeApply
-	opReq.View = views.NewOperation(arguments.ViewHuman, c.RunningInAutomation, c.View)
-
-	opReq.ConfigLoader, err = c.initConfigLoader()
-	if err != nil {
-		c.showDiagnostics(err)
-		return 1
+	opReq.View = view.Operation()
+	// FIXME: this shim is needed until the remote backend is migrated to views
+	opReq.ShowDiagnostics = func(vals ...interface{}) {
+		var diags tfdiags.Diagnostics
+		diags = diags.Append(vals...)
+		view.Diagnostics(diags)
 	}
 
-	{
-		var moreDiags tfdiags.Diagnostics
-		opReq.Variables, moreDiags = c.collectVariableValues()
-		diags = diags.Append(moreDiags)
-		if moreDiags.HasErrors() {
-			c.showDiagnostics(diags)
-			return 1
-		}
-	}
-
-	op, err := c.RunOperation(be, opReq)
+	var err error
+	opReq.ConfigLoader, err = c.initConfigLoader()
 	if err != nil {
-		c.showDiagnostics(err)
-		return 1
-	}
-
-	if op.Result != backend.OperationSuccess {
-		return op.Result.ExitStatus()
+		diags = diags.Append(fmt.Errorf("Failed to initialize config loader: %s", err))
+		return nil, diags
 	}
 
-	// Show the count results from the operation
-	if c.Destroy {
-		c.Ui.Output(c.Colorize().Color(fmt.Sprintf(
-			"[reset][bold][green]\n"+
-				"Destroy complete! Resources: %d destroyed.",
-			countHook.Removed)))
-	} else {
-		c.Ui.Output(c.Colorize().Color(fmt.Sprintf(
-			"[reset][bold][green]\n"+
-				"Apply complete! Resources: %d added, %d changed, %d destroyed.",
-			countHook.Added,
-			countHook.Changed,
-			countHook.Removed)))
-	}
-
-	// only show the state file help message if the state is local.
-	if (countHook.Added > 0 || countHook.Changed > 0) && c.Meta.stateOutPath != "" {
-		c.Ui.Output(c.Colorize().Color(fmt.Sprintf(
-			"[reset]\n"+
-				"The state of your infrastructure has been saved to the path\n"+
-				"below. This state is required to modify and destroy your\n"+
-				"infrastructure, so keep it safe. To inspect the complete state\n"+
-				"use the `terraform show` command.\n\n"+
-				"State path: %s",
-			c.Meta.stateOutPath)))
-	}
```

Comment:
🚨 **CONFUSION** 🚨

I can't find any way to reach this code path. Even setting the `-state-out=` flag and adding some resources does not result in this being displayed. I've tested this on Terraform versions going back to 0.12.29.

As a result, I've removed it here. I'd be happy to add it back if there is a way to reach it that I'm missing.

Reply:
I'm able to get here on a fresh build from master: 

```
random_pet.meow[2]: Creation complete after 0s [id=in-pup]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: local.tfstate
```

Reply:
Will fix this, thank you! I have no idea why I had such trouble reproducing it before.

Reply:
Fixed in the latest force push:

<img width="1019" alt="state-out" src="https://user-images.githubusercontent.com/68917/108736291-77c7d700-74ff-11eb-9b65-322cf127792b.png">

Also, I think I understand why I couldn't reproduce this before: I cannot reach this code on any prior release of Terraform, only on builds from current master or this branch. Somewhat concerning… but maybe we accidentally fixed a bug recently?

Reply:
🤷‍♀️ what is a terraform, even?